### ISSUE TITLE
How to set hyperparameters for speaker diarization pipeline?

### ISSUE THREAD

**Issue by sunraymoonbeam** [pyannote/pyannote-audio#1579]
I am currently working on a speaker diarization task for classroom discussions without labeled data. To assess the pipeline's performance, I rely on two methods: listening manually and using intrinsic measures such as clustering metrics. The pipeline, when used out of the box, doesn't perform well. Some segments contain background noise, while others are very short (e.g., 0.1 seconds).

I want to improve the diarization pipeline's performance by tweaking hyperparameters. I know about the hyperparameters for segmentation (_threshold_, _min_duration_off_, and _min_duration_on_) and clustering (_method_, _min_cluster_size_, and _threshold_). However, I'm having trouble instantiating hyperparameters for the segmentation model.

Here's my attempt:

```
pretrained_pipeline = Pipeline.from_pretrained("pyannote/speaker-diarization-3.1", use_auth_token=True)
default_hyperparameters = pretrained_pipeline.parameters(instantiated=True)
for param, value in default_hyperparameters.items():
    print(f"{param}: {value}")
```
![Screenshot 2023-12-04 at 17 36 19](https://github.com/pyannote/pyannote-audio/assets/98078901/509a2a22-3e8d-4d3c-923f-c643ca4d4bc2)

The output only shows one tunable hyperparameter for segmentation (_min_duration_off_). After some investigation, I discovered that using **_pyannote/segmentation-3.0_** for the segmentation model results in only _min_duration_off_ being visible. However, when using `pyannote.audio.pipelines.SpeakerDiarization` pipeline with the default segmentation model **_pyannote/segmentation@2022.07_**, the _threshold_ activation parameter is available.

![Screenshot 2023-12-04 at 17 35 30](https://github.com/pyannote/pyannote-audio/assets/98078901/cc435f7d-f90f-459e-832a-496dcd4897ea)

I'm curious about the difference between these segmentation models. Additionally, I noticed that the VAD pipeline has _min_duration_on_, but the speaker diarization pipeline does not (which I would like to remove those short speech segments). Initially, I performed each task separately (VAD -> Embedding of speech segments -> Clustering) instead of using the pipeline. My understanding is that the VAD pipeline doesn't account for speaker change detection and only detects regions of speech, which is why I switched back to the pipeline for easy inference and testing.

**### Questions**
1. What is the difference between the segmentation models and why do they have different settable hyper-parameters?
2. Is it possible to set "min_duration_on" for the speaker diarization pipeline?
3. What is the difference between VAD pipeline and how the segmentation model works in the Speaker Diarization Pipeline? My understanding is that the VAD pipeline uses the maximum over the speaker axis for each frame, but the Segmentation model performs speaker change detection by taking the absolute value of the first derivative over the time axis, and take the maximum value over the speaker axis.

Can you provide insights into these issues and advice me on how to proceed for playing around with the hyperparameters to improve my performance?

Warm regards,
Zack

**Comment by hbredin** [comment]
> * What is the difference between the segmentation models and why do they have different settable hyper-parameters?

`pyannote/segmentation` is described in [this paper](https://www.isca-speech.org/archive/pdfs/interspeech_2021/bredin21_interspeech.pdf).
`pyannote/segmentation-3.0` is described in [that paper](https://www.isca-speech.org/archive/pdfs/interspeech_2023/plaquet23_interspeech.pdf).

This should help you understand why the latter does not need for `onset` and `offset` thresholds.

> * Is it possible to set "min_duration_on" for the speaker diarization pipeline?

Not out of the box. You would have to post process the output yourself.
Pipelines output `pyannote.audio.Annotation` instances so [this API](http://pyannote.github.io/pyannote-core/reference.html#annotation) might help you do just that.

> * What is the difference between VAD pipeline and how the segmentation model works in the Speaker Diarization Pipeline? 
> * My understanding is that the VAD pipeline uses the maximum over the speaker axis for each frame

Correct. Plus thresholding.

> * but the Segmentation model performs speaker change detection by taking the absolute value of the first derivative over the time axis, and take the maximum value over the speaker axis.

Incorrect. [See this paper](https://www.isca-speech.org/archive/pdfs/interspeech_2023/bredin23_interspeech.pdf) explaining how the segmentation model is used.

**Comment by picheny-nyu** [comment]
Does this mean that there is no need to tune the segmentation hyperparameters for 3.0 so this step should be omitted?

**Comment by avion23** [comment]
Here's a Python snippet I used with `pyannote/speaker-diarization-3.1`:

```python
from pyannote.pipeline import Pipeline

def get_pyannote_segments(audio_file):
    pipeline = Pipeline.from_pretrained("pyannote/speaker-diarization-3.1")

    pipeline.parameters(segmentation={'min_duration_off': 0.0})
    pipeline.parameters(clustering={
                        'method': 'centroid', 'min_cluster_size': 12, 'threshold': 0.7045654963945799})

    diarization = pipeline({'audio': audio_file})
    return [({"start": segment.start, "end": segment.end}, label) for segment, _, label in diarization.itertracks(yield_label=True)]
  ```

**Comment by avion23** [comment]
pyannote/speaker-diarization-3.1.1 has more parameters, but it turns out they are read only :( 

Probably that's only documentation on why the model was trained.

**Comment by NorthernNoober** [comment]
Are there any hyper parameters for any available models that I can use to identify distinct voice chunks in my file?  I am passing to my script the speakers and the order they speak in an array [Speaker1, Speaker2, Speaker1, Speaker3] and I have ensured there is exactly 1 second of silence between each speaker (which I can make longer if need be).  My hope is to get the model to identify the correct number of turns since I already know the identities and I have my script adding the correct speaker identities into the JSON.  I would like to then have it not parse on any silence less than 1 second to ensure this happens properly.  The issue I'm having is if I have three speakers speaking once each, often it sees four turns or more and fails because of a pause in one of the speakers dialog.

OBE:  I just had my audio separate speakers by at least two seconds, let this model produce its JSON with its accurate chunking of conversation but inaccurate identification of speakers (they're AI generated voices so maybe that's why it can't distinguish) and then I coded something else to fix the JSON and take the whitespace out.
